src/optimize/scalar/bracket.rs

```diff
@@ -0,0 +1,188 @@
+use std::cmp::Ordering::Equal;
+use std::fmt::Debug;
+use num_traits::{abs, Float, FromPrimitive, Pow, Signed};
+use thiserror::Error;
+
+#[derive(Debug, Copy, Clone)]
+pub struct Bracket<T: Float + PartialOrd + Debug> {
+	pub left: T,
+	pub f_left: T,
+	pub right: T,
+	pub f_right: T,
+	pub center: T,
+	pub f_center: T
+}
+
+#[derive(Error, Debug, Copy, Clone)]
+pub enum BracketError{
+	#[error("A bracket bound is duplicated and therefore a bracket would be invalid")]
+	DupeBoundForBracket,
+	#[error("The specified bracket does not encapsulate a dip. It might have a local minima within the bound but this is unknown")]
+	BracketNotADip
+}
+
+
+impl<T: Float + PartialOrd + Debug> Bracket<T>{
+	pub fn new(x1: T, x2: T, x3: T, func: fn(T) -> T) -> Result<Self, BracketError>{
+		let mut elements: [T; 3] = [x1, x2, x3];
+
+		elements.sort_by(|a, b| a.partial_cmp(b).unwrap_or(Equal));
+
+		Self::new_processed(elements[0], func(elements[0]), elements[1], func(elements[1]), elements[2], func(elements[2]))
+	}
+
+	pub fn new_processed(left: T, f_left: T, center: T, f_center: T, right: T, f_right: T) -> Result<Self, BracketError>{
+
+		if left == center || center == right{
+			return Err(BracketError::DupeBoundForBracket)
```

Comment:
maybe provide in the error which bound is duplicated?

Reply:
Fixed in 76af062c2be80746199488acc19ab0bc0068c2d3.